cirq-core/cirq/ops/dense_pauli_string.py

```diff
@@ -188,84 +185,89 @@ def __pow__(self, power):
 
         raise TypeError(f'indices must be integers or slices, not {type(item)}')
 
     def __iter__(self) -> Iterator['cirq.Gate']:
         for i in range(len(self)):
             yield self[i]
 
-    def __len__(self):
+    def __len__(self) -> int:
         return len(self.pauli_mask)
 
-    def __neg__(self):
-        return DensePauliString(coefficient=-self.coefficient, pauli_mask=self.pauli_mask)
+    def __neg__(self: T) -> T:  # type: ignore
+        cls = type(self)
+        return cls(coefficient=-self.coefficient, pauli_mask=self.pauli_mask)
 
-    def __truediv__(self, other):
-        if isinstance(other, (sympy.Basic, numbers.Number)):
+    def __truediv__(self: T, other: Any) -> T:  # type: ignore
+        if isinstance(other, (sympy.Basic, int, float, complex)):
             return self.__mul__(1 / other)
 
         return NotImplemented
 
-    def __mul__(self, other):
-        if isinstance(other, BaseDensePauliString):
-            max_len = max(len(self.pauli_mask), len(other.pauli_mask))
-            min_len = min(len(self.pauli_mask), len(other.pauli_mask))
+    def __mul__(self: T, other: Any) -> T:  # type: ignore
+        cls = type(self)
+
+        if isinstance(other, cls):
+            max_len = max(len(self), len(other))
+            min_len = min(len(self), len(other))
             new_mask = np.zeros(max_len, dtype=np.uint8)
-            new_mask[: len(self.pauli_mask)] ^= self.pauli_mask
-            new_mask[: len(other.pauli_mask)] ^= other.pauli_mask
+            new_mask[: len(self)] ^= self.pauli_mask
+            new_mask[: len(other)] ^= other.pauli_mask
             tweak = _vectorized_pauli_mul_phase(
                 self.pauli_mask[:min_len], other.pauli_mask[:min_len]
             )
-            return DensePauliString(
+            return cls(
                 pauli_mask=new_mask, coefficient=self.coefficient * other.coefficient * tweak
             )
 
-        if isinstance(other, (sympy.Basic, numbers.Number)):
+        if isinstance(other, (sympy.Basic, int, float, complex)):
             new_coef = protocols.mul(self.coefficient, other, default=None)
             if new_coef is None:
                 return NotImplemented
-            return DensePauliString(pauli_mask=self.pauli_mask, coefficient=new_coef)
+            return cls(pauli_mask=self.pauli_mask, coefficient=new_coef)
 
         split = _attempt_value_to_pauli_index(other)
         if split is not None:
             p, i = split
             mask = np.copy(self.pauli_mask)
             mask[i] ^= p
-            return DensePauliString(
+            return cls(
                 pauli_mask=mask,
                 coefficient=self.coefficient * _vectorized_pauli_mul_phase(self.pauli_mask[i], p),
             )
 
         return NotImplemented
 
-    def __rmul__(self, other):
-        if isinstance(other, (sympy.Basic, numbers.Number)):
+    def __rmul__(self: T, other: Any) -> T:  # type: ignore
```

Comment:
Why the type ignore? (here and elsewhere). Should we revert to not having type annotations in these cases instead of explicitly ignoring them?

Reply:
The thing here is that `BaseDensePauliString` is a subclass of `raw_types.Gate`. As a result, mypy expects `BaseDensePauliString.__rmul__` to return an object of a subclass of `cirq.LinearCombinationOfGates`. That is not the case, as the operation actually acts on the `coefficient` attribute of the Pauli string. Maybe a natural fix would be for the operations in `raw_types.Gate` to have `Union[Gate, cirq.LinearCombinationOfGates]` as return type instead of `cirq.LinearCombinationOfGates` alone. I'll commit the fix for review.

Reply:
I rollbacked the previous change, that is, the return type of linear operations  in `raw_types.py` is now back to `cirq.LinearCombinationOfGates` alone. I used `Gate`'s method `wrap_in_linear_combination` to convert the results of linear operations in `dense_pauli_string.py` (I mimicked what's done in `raw_types.py`).